Optimization changes to completely fold OptionSet literals

eeckstein opened the pull request:
With this the optimizer is able to fold OptionSet literals like
     [.first, .second, .third, .fourth]
to a single integer constant. See test/SILOptimizer/optionset.swift

This PR consists of a couple of commits, which are (mostly) small changes to existing optimizations:

* DeadObjectElimination: handle objects for which the destructor is de-virtualized and inlined
* Tuning the inliner heuristics
* ARCCodeMotion: add a small peephole optimization which removes adjacent retain/release pairs.
* GlobalOpt: Move the object outlining from GlobalOpt into a separate pass "ObjectOutliner"
* SideEffectAnalysis: don't assume the worst side-effects for a release instruction
* SideEffectAnalysis: handle unchecked_ref_cast instructions
* SimplifyCFG: do constant folding also in SimplifyCFG.
* Move the constant folding logic into a separate utility which can be used by multiple passes.

For details see the commit messages

rdar://problem/26411312
and also rdar://problem/34758773

eeckstein committed:
Move the constant folding logic into a separate utility which can be used by multiple passes.

NFC

eeckstein committed:
SimplifyCFG: do constant folding also in SimplifyCFG.

This enables to optimize chains of constant-foldable terminal instructions, for which we would have needed multiple ConstantPropagaion-SimplifyCFG pass pairs in the pipeline otherwise.
See the test file for an example.

eeckstein committed:
SideEffectAnalysis: handle unchecked_ref_cast instructions

eeckstein committed:
SideEffectAnalysis: don't assume the worst side-effects for a release instruction

Instead let the client decide what to do with this.
Sometimes the client knows what side effect a release instruction really has.

eeckstein committed:
GlobalOpt: Move the object outlining from GlobalOpt into a separate pass "ObjectOutliner"

We run GlobalOpt multiple times in the pass pipeline but in some cases object outlining shouldn't be done too early.
Having it done in a separate pass enables to run it independently from GlobalOpt.

eeckstein committed:
ARCCodeMotion: add a small peephole optimization which removes adjacent retain/release pairs.

This reduces phase ordering problems, because it eliminates the need for a ARCSequenceOpt (which also does this optimization).

eeckstein committed:
Tuning the inliner heuristics:

* allow small class methods to be inlined with -Osize
* inline pure calls: references to objects, which are initialized with constants, are considered as constant arguments

eeckstein committed:
DeadObjectElimination: handle objects for which the destructor is de-virtualized and inlined

Also: add an additional DeadObjectElimination pass in the low level pipeline because
redundant load elimination (which runs before) can turn an object into a dead object.

eeckstein committed:
Add a test to check if OptionSet literals are fully folded

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci smoke benchmark

eeckstein commented:
@swift-ci Please test compiler performance

jrose-apple commented on lib/SILOptimizer/Analysis/SideEffectAnalysis.cpp:
```diff
@@ -347,10 +347,6 @@ void SideEffectAnalysis::analyzeInstruction(FunctionInfo *FInfo,
-      
-      // TODO: Check the call graph to be less conservative about what
-      // destructors might be called.
-      FInfo->FE.setWorstEffects();
```
I don't see how this is safe in general. Shouldn't the client have to say explicitly when they know a value hasn't escaped?

eeckstein replied:
FunctionEffects::getMemBehavior() returns the worst effects anyway if one of the parameter effects contains a "release".
The only thing which changed is that if a client explicitly looks at parameter effects, the client has to consider that a "released" parameter also potentially means worst effects on other parameters.

jrose-apple replied:
Hm. It feels like it's too easy for someone to forget that, but I'll have to trust that the design makes sense for SideEffectAnalysis.

eeckstein replied:
Well, I should add a comment to the ABI functions

eeckstein commented:
@swift-ci Please test compiler performance

eeckstein committed:
stdlib: make the destructor of _SwiftNativeNSArrayWithContiguousStorage inlinable on linux

... as it is for ObjC runtimes.
This is need for the optimizer to do certain optimizations, like object outlining.

rdar://problem/34758773

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci Please smoke test compiler performance

eeckstein commented:
@swift-ci Please smoke test compiler performance

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci Please smoke test compiler performance

eeckstein commented:
@swift-ci Please smoke test compiler performance

shahmishal commented:
@swift-ci Please smoke test compiler performance

shahmishal commented:
@swift-ci Please smoke test compiler performance

eeckstein commented:
@swift-ci smoke test and merge

eeckstein commented:
@swift-ci smoke test and merge
